lfscs-meetings/resources/Minimum_Linux_Requirements_for_Executing_a_Minimal_Application.md

```diff
@@ -0,0 +1,455 @@
+
+Tracing kernel execution is not always a complete process. Some functions are 
+explicitly marked as non-traceable using the `noinstr` attribute, preventing 
+them from appearing in ftrace logs. Additionally, compiler optimizations can 
+exclude function calls from tracing, creating gaps in the recorded execution 
+path.
+
+To assess these missing traces, we compare the kernel’s symbol table against 
+the list of traceable functions:
+
+```
+cat /proc/kallsyms | cut -d " " -f3 | sort > all
+sudo cat /sys/kernel/tracing/available_filter_functions | cut -d " " -f1 | sort > avail
+vimdiff all avail
+```
+
+This analysis reveals that numerous functions remain untraceable, 
+necessitating alternative methods to infer missing data.
+
+### **Filling the Gaps with Static Analysis**
+
+To compensate for missing function calls in our ftrace log, we leverage 
+`ks-nav`’s static analysis capabilities. By overlaying a static function call 
+graph onto the dynamically traced execution flow, we can infer missing 
+connections. The complete function of `graph-tool` enables us to merge 
+these datasets, visually distinguishing inferred connections.
+
+For example, integrating a static graph of `execve` with the ftrace-derived 
+graph introduces additional inferred edges:
+
+```
+./graph-tool complete -f1 executed.dot -f2 execve.dot -color green | xdot -
+```
+
+While this approach improves completeness, it is not foolproof... Some 
+indirect calls, such as those involving function pointers, remain unresolved.
+
+### **Dealing with Indirect Calls**
+
+Kernel execution frequently involves indirect function calls, where function 
+pointers dynamically determine the next function to execute. 
+These calls obscure direct function transitions, complicating execution graph 
+reconstruction.
+
+For instance, the function `make_prot` may be identified as part of 
+`load_elf_binary`, yet its trace lacks a direct link from its actual caller, 
+`bprm_execve`. To resolve this, we query ks-nav’s function call database:
+
+```
+ssh dbs "psql -d kernel_bin -c \"select * from symbols where symbol_id in (select caller from xrefs where callee in (select symbol_id from symbols where symbol_name = 'make_prot' and symbol_instance_id_ref=56));\""
+```
+
+This enables us to manually integrate missing links by generating a static 
+function call graph for `load_elf_binary` and connecting it to our existing 
+graph:
+
+```
+./graph-tool connect -f1 execve.dot -f2 load_elf_binary.dot -n1 bprm_execve -n2 load_elf_binary > execve_load_elf_binary.dot
+./graph-tool complete -f1 executed.dot -f2 execve_load_elf_binary.dot -color green | xdot -
+```
+
+Through these steps, we refine our understanding of kernel execution, 
+compensating for the limitations of dynamic tracing alone. 
+This combined approach allows us to construct a more complete and accurate 
+representation of the minimal kernel requirements for executing a basic 
+application.
+
```

Comment:
And also here it would be good to show the graph pre and post indirect call compensation